### ISSUE TITLE
extended_info

### ISSUE THREAD

**Issue by AnthonyHunt** [leihs/leihs#93]
Hi All,

I used to write our hashed card numbers and barcodes into the extended_info field, and oddly need to have the badge_id there too for photos to work.  This used to be stored as just plain text, but this now crashes the edit button with: 

ActionView::Template::Error (undefined method `each' for "048055048048055049048048049000 212821212822212122 070071001":String):
    2: 
    3: .row.margin-vertical-s
    4: 
    5:   - @user.extended_info.each do |k, v|
    6:     .row.padding-inset-s.emboss.margin-bottom-xxs
    7:       .col1of2
    8:         %strong.font-size-m= "#{k}"
  app/views/manage/users/_extended_info.html.haml:5:in `_app_views_manage_users__extended_info_html_haml___195879586751455574_20785040'

So, my question is... how should I now write these values to this field?  I currently update this field at login in the HSLU_authenication_coltroller with:

    user.badge_id = user_data["extensionattribute5"].first.to_s unless user_data["extensionattribute5"].first.blank?
    user.extended_info = user_data["extensionattribute5"].first.to_s unless user_data["extensionattribute5"].first.blank?

    tmp_y = user.extended_info.unpack('C*').collect{|a| a - 27}.join unless user.extended_info.blank?
    user.extended_info = "0" + user.extended_info.unpack('C*').join("0") + "000" unless user.extended_info.blank?
    user.extended_info = user.extended_info + " " + tmp_y + " " + user_data["extensionattribute5"].first.to_s unless user.extended_info.blank?

Thanks,

Anthony.

**Comment by psy-q** [comment]
extended_info was always expected to store a serialized hash, not a string, so if you write anything into it as a hash, it should just show up. For example:

```ruby
user.extended_info = { favorite_color: 'brown',
                             pet_type: 'dog' }
user.save
```

And that information will now show up on the user edit page. The reason it explodes now for you if you're written plain strings into that field might be that only now have we started displaying all the extended info on the user page. It might seem like the database should take a string there, since the column type is text, but text is also the type ActiveRecord uses for any serialized Ruby data.

`badge_id` on the other hand is its own, real column on the database and every user has one, there you can write just a plain string.

Then there's `unique_id` which should contain a unique ID of the user, as saved in some external system (LDAP, CAS, anything). It can be interpolated into the URL defined in settings under  user_image_url, replacing the `{:id}` placeholder, if you want to show user portraits in the leihs interface.

So if you have a user_image_url URL like `http://example.org/students/image?id={:id}` that spits out user portraits of the right size, that should just work.

What else are you using the barcode information for? Could you just store it in `badge_id`? It will be found when scanned, if what you're doing is scanning user ID cards with that barcode on them.

You can also save any other information in extended_info, just as a hash, not as a string. The only place the leihs code itself uses `extended_info` is on the user edit page (showing all the information with each hash key as a title, each value as a value) and when deciding what to interpolate into the `user_image_url`.

If the user has a `unique_id` and `user_image_url` has a `{:id}` in it, `{:id}` will be replaced with the user's `unique_id`.

If the user does not have a `unique_id` but does have an `extended_info[:id]`, any occurrence of `{:extended_info:id}` in `user_image_url` will be replaced with the user's `extended_info[:id]`.

**Comment by AnthonyHunt** [comment]
Hi,

Thanks for the reply.  Our system is probably over complicated but I'll explain how our fields are assigned.

Each user has their own Login ID for our AD network.  This is in the unique_id field, and the login field. If it's not in unique_id they can't log in.  We use badge_id to store a special value which is the students student number and issue number of the card, this is a tie up from v2 where it was a search field.

The two other things we need to store (and must be able to search for is the encoded card number (a unique id from their student card.  If they loose their card, a new card is issued with a new id). The encoded value of the barcode on their card.  Both of these values are needed as we can scan they card with either a card reader or barcode reader.

The last thing we store is the students student number.  This is unique to the student and is how the students photo is stored.

So, to sum up: we need to be able to search for user barcode, user card code or card number.  Previously I had modified app/models/user.rb to include this field.

                      or(arel_table[:extended_info].matches(q)).

Might I be better putting these other value inside the badge_id field now?

Thanks,

Ant.

**Comment by AnthonyHunt** [comment]
Ok, so I've been refactoring my authentication controllor to taking into account the above. I've got it all working. I've put my card code, barcode and (student number + issue number) into my badge_id. Searches seem to be working ok.

For good measure I've even added them into extended_info and they show up properly in the users details. 

I have however found a problem. I'm now storing the student number (that gives the photo id) in extended_info[id].  This is strictly numeric but written to the array as attribute["student number"].first.to_s. It appears as a string as far as I can tell. When I use this in the image url it appears to appear as a url encoded string ("%7c" or something similar (not sat at computer)). Looking at app/models/user.rb I can see the unique_id lines doesn't have a .to_s where the extended_info:id entry does. I'm also not sure the "and extended_info and extended_info[id]" logic is working correctly? By standard I get no image, but if I edit the unique_id code above it I at least get a placeholder (with the garbage image id).

Make sense?

**Comment by AnthonyHunt** [comment]
ok, so I've been doing more testing today.  I swapped out the unique_id gsub with extended_info['id'] to force it to replace with the extended_info field. I just get an empty field.  So, as far as I can tell, the user.rb doesn't read the extended_info field? is that really true?

**Comment by psy-q** [comment]
The User model does use extended_info, but only for two things:

 1. An alternate email address on `user.alternative_email` if `extended_info['email_alt']` is filled in. This alternate email address is used in reminders, confirmations etc. to make sure messages reach users and they're less likely to say things like "but I never check my university email account".
 2. user.image_url, which is constructed from interpolating `extended_info['id']` into the URL configured in settings under `user_image_url`, replacing any occurrence of the placeholder {:extended_info:id}

I actually gave a symbol `extended_info[:id]` as a key for the hash in the example above, but it should really be a string: `extended_info['id']`. Sorry if that's the cause of the confusion. I'll amend the comment above.

As long as extended_info is a hash of strings pointing to whatever (usually also strings), I think you shouldn't run into any trouble. I can't explain the garbage in the URL, but I'll try to reproduce that here. The .to_s should actually not even be necessary if you ask me, since I think `String#gsub` should already cast the replacement into a string, but I'm not sure there either.

**Comment by AnthonyHunt** [comment]
Hmm, that's what I have, and it's not returning anything. If I remove the .to_s it crashes. All the elements show up properly in the users details page though. It's just as if they aren't bring read by user.rb? Very odd behaviour.

**Comment by AnthonyHunt** [comment]
Ok.  It's more fundamental than that, and I have no idea how it has happened.  I thought some thing must else must be causing a problem so I compared my user.rb with the master.  Mine has Setting::USER_IMAGE_URL in uppercase, yours is Setting.user_image_url in mixed case.  Yours has 317 lines in it, mine 340.  Now, I have only made one change to that file after my initial fork which was to add the extended_info into the search sql.  If I replace my file with the master it breaks LEIHS based on the case issue above.  So.... this means, that somehow even though my fork is pulled from your and merged with only a couple of changes.... we are out of sync. Darn! 

I'm guessing that there's some changes further up that are causing this issue to arise.  I've just trying to find the changes (i think it's only this change and the hslu controller) and I'll have to zap the fork and start over. :(

**Comment by psy-q** [comment]
Don't worry about `Setting.user_image_url`, while the settings are stored as constants (e.g. `Setting::USER_IMAGE_URL`), there is a method to access them so that they map to e.g. `Setting.user_image_url`. But it could be that if your Setting model differs, there could be trouble accessing them.

Your fork might have diverged in some other ways -- I guess the safest way to track our master is to keep a separate remote (in git terms) that you can pull from, so that you can merge our master into it from there from time to time:

```
git remote add zhdk https://github.com/zhdk/leihs.git
git checkout master
```
(do some changes to your own master here)
```
git merge zhdk master
```

That should merge your changes with our master and it will conflict when there are clashes. That way git would let you know exactly where you have to take action to bring your changes in.

That the .to_s should be necessary is odd, since what I think you're doing is starting with strings and concatenating to them anyhow, so they should remain strings the whole way through.

I would love to suggest that we do a more configurable login system for plain old vanilla leihs instead, so that you wouldn't have to customize that, but the behavior you need is so specific that we'd end up having to do weird insecure and crazy things like allowing code inside the settings, so for now I'm afraid this seems to be the best way.

**Comment by AnthonyHunt** [comment]
Hmm, ok.  Well it should be working like this already.  I forked LEIHS a year or so ago.  My process is almost what you have above.  I have only 2 changes nowadays which is to:

	app/controllers/authenticator/hslu_authentication_controller.rb
	app/models/user.rb

My process on my test server is:

git fetch upstream 
git merge  3.9.0
git push origin master
rvm use 2.1.1
bundle install --deployment --without cucumber development
RAILS_ENV=production bundle exec rake db:migrate
RAILS_ENV=production bundle exec rake assets:precompile
touch tmp/restart.txt


and on my production server:

git fetch --all
git reset --hard origin/master
rvm use 2.1.1
bundle install --deployment --without cucumber development
RAILS_ENV=production bundle exec rake db:migrate
RAILS_ENV=production bundle exec rake assets:precompile
touch tmp/restart.txt


So I'm slightly bamboozled to why I now have these problems?

Ant.

**Comment by psy-q** [comment]
Hm, if the Setting model isn't different, things should just work as written above. If your newer pulls from our master merge fine with your work except for perhaps the two conflicts that you expect, I don't know what the issue could be. I will poke @sellittf if he has an idea where to look.

Is your modified version the one at AnthonyHunt/leihs? We can diff it to ours and sift through things a little. I'm hopeful it's just a tiny thing that's easily missed.

**Comment by AnthonyHunt** [comment]
As far as I remember from last week, it was just a couple of case things.  Yes, that's where it's sitting.  I tried the diff in GitHub, but it gives me about 7,000 changes because for some reason the rubies are getting sync'd even though they are ignored... :(.  So glad it's summer and I can look at a few of these things.

**Comment by psy-q** [comment]
Hmm, it looks like the git histories have diverged a bit. It seems most of the Rubocop-based cleanup work we've done didn't trickle down to your fork either.

Without investing more time it's hard to suggest the best thing to do, but what happens if you clone one of our stable tags all plain and vanilla, make your small changes and run that on a test server just to see? If that works, maybe the fastest way to get back on track could be redoing your changes on top of a plain new clone from our master.

I'd also be interested in looking at what changes exactly are necessary for you. I will see if we could add extended_info the global search without too many unwanted side effects. That would leave you with just one thing to take care of.

**Comment by AnthonyHunt** [comment]
Yes, thoroughly confusing to why it's happening.  I've considered rebuilding from scratch and that seems the most viable option at the moment.  The only other change is the LDAP/HSLU authentication controller that adds our specific search terms into extended_info and badge_id.

I'm hopeful that if I copy out the controller, LDAP.yml and Database.yml, and then the attachments folder, I should be able to bin the LEIHS directory, bin my fork, re-fork, replace files and push it back to my fork without any issues.... but I normally get some just to spice it up!

**Comment by AnthonyHunt** [comment]
Rebuilt my fork today from scratch.  Most things are working as they should have been.  Still not getting photos to appear when using :extended_info:id as the source.  I think there's a logic issue with how it gets the string from :extended_info:id? and converts it to a string?  It seems to fail it logic test on "elsif Setting.user_image_url.match(/\{:extended_info:id\}/) and extended_info and extended_info['id']"

Ant.

**Comment by psy-q** [comment]
Do the users really have `extended_info[:id]` set? I know it might be a stupid question. Here's the logic:

```ruby
  def image_url
    if Setting.user_image_url
      if Setting.user_image_url.match(/\{:id\}/) and unique_id
        Setting.user_image_url.gsub(/\{:id\}/, unique_id)
      elsif Setting.user_image_url.match(/\{:extended_info:id\}/) and extended_info and extended_info['id']
        Setting.user_image_url.gsub(/\{:extended_info:id\}/, extended_info['id'].to_s)
      end
    end
  end
```
So if there is a setting for `user_image_url` and if that particular user has a `unique_id`, replace any occurence of `{:id}` in the `user_image_url` with that user's `unique_id`.

If none of the before is true: if `user_image_url` contains `{:extended_info:id}` **and** the user has `extended_info` **and** the user has `extended_info['id']`, then replace any occurence of `{:extended_info:id}` in `user_image_url` with that user's `extended_info['id']`.

If you go into a Rails console and retrieve a particular user that you know has complete data, what are the results for that one? For example, what's the output of each of these, replacing 1234 with the user ID of that user:

```
bundle exec rails c
Setting.user_image_url
u = User.find(12345)
u.extended_info[:id]
u.image_url
```

**Comment by AnthonyHunt** [comment]
The users have that info and it's in the database:

![image](https://cloud.githubusercontent.com/assets/299919/10754771/cf0ed1ec-7c91-11e5-8328-8736abea2352.png)

I'm just going to check the console, but the info is read with other pages:

![image](https://cloud.githubusercontent.com/assets/299919/10754828/2ac01294-7c92-11e5-8a0c-4d216c2dac25.png)


Ant.

**Comment by AnthonyHunt** [comment]
Here's the output:

2.1.1 :003 > u = User.find(1714)
  User Load (0.3ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1714 LIMIT 1
 => #<User id: 1714, login: "GRET1_15", firstname: "Terry", lastname: "Greer", phone: nil, authentication_system_id: 4, unique_id: "GRET1_15", email: "gret1_15@***.ac.uk", created_at: "2015-10-23 13:55:56", updated_at: "2015-10-23 13:55:56", badge_id: "15015354", address: nil, city: nil, zip: nil, country: nil, language_id: 2, extended_info: {:mifare=>"222621222624262522", :barcode=>"049053048049053051053052049000", :id=>"15015354"}, settings: {}, delegator_user_id: nil> 
2.1.1 :004 > u.extended_info[:id]
 => "15015354" 
2.1.1 :005 > u.image_url
 => nil

**Comment by psy-q** [comment]
Yes, it looks like the logic really isn't matching. What is the output of `Setting.user_image_url`?

**Comment by AnthonyHunt** [comment]
That's correct with:

2.1.1 :001 > Setting.user_image_url
 => "https://secure.worc.ac.uk/registry/stu_cards/stu_images/{:extended_info:id}.jpg" 

if I change it to just :id i get a small square image placeholder (because is getting a 404 on the image) but with :extended_info:id set it doesn't even get anything.  If I remove the logic checks, I obviously get the nil image (which suggests to me that user.rb can't read the array properly?

Just a side note for when it does work (as I've previously cheated and hardcoded it) can the image be set to scale properly? like it does in the toolbar? The images that are stored are all sorts of sizes (not done by us) and it would be great if they could be set to a  size.

**Comment by psy-q** [comment]
Ah, finally something to grasp: The `user_image` method uses the `id` string as a key for looking up extended_info ID's, not the `:id` symbol. So it looks for `extended_info['id']` but what's stored on your user is `extended_info[:id]`. Ruby hashes can use anything as keys, so those are two different keys.

The ZHdK authentication controllers maps directly from XML to extended_info keys, that's why it sets strings as keys (since the parsed XML can't be anything other than strings anyhow).

Can you modify your authentication controller to save to `extended_info['id']` instead of `extended_info[:id]`?

**Comment by AnthonyHunt** [comment]
So would that go from:

    user.extended_info = { mifare: (user_data["extensionattribute5"].first.to_s.unpack('C*').collect{|a| a - 27}.join unless user_data["extensionattribute5"].first.blank?),
                             barcode: ("0" + user_data["extensionattribute5"].first.to_s.unpack('C*').join("0") + "000" unless user_data["extensionattribute5"].first.blank?),
                              id: (user_data["extensionattribute1"].first.to_s),
                             modules: (user_data["extensionattribute7"].first.to_s unless user_data["extensionattribute7"].first.blank?) }

to:

    user.extended_info = { 'mifare': (user_data["extensionattribute5"].first.to_s.unpack('C*').collect{|a| a - 27}.join unless user_data["extensionattribute5"].first.blank?),
                             'barcode': ("0" + user_data["extensionattribute5"].first.to_s.unpack('C*').join("0") + "000" unless user_data["extensionattribute5"].first.blank?),
                              'id': (user_data["extensionattribute1"].first.to_s),
                             'modules': (user_data["extensionattribute7"].first.to_s unless user_data["extensionattribute7"].first.blank?) }



Thanks.

**Comment by AnthonyHunt** [comment]
That appears to have sorted the problem.

THANKS!!!!!!!!!

**Comment by AnthonyHunt** [comment]
So I start testing this.  I replace (commented out) my original code with this:

    user.extended_info['id'] = (user_data["extensionattribute1"].first.to_s unless user_data["extensionattribute1"].first.blank?)
    user.extended_info['mifare'] = (user_data["extensionattribute5"].first.to_s.unpack('C*').collect{|a| a - 27}.join unless user_data["extensionattribute5"].first.blank?)
    user.extended_info['barcode'] = ("0" + user_data["extensionattribute5"].first.to_s.unpack('C*').join("0") + "000" unless user_data["extensionattribute5"].first.blank?)
    user.extended_info['modules'] = (user_data["extensionattribute7"].first.to_s unless user_data["extensionattribute7"].first.blank?)

This gave me duplication of the fields (as I expected) in the users screen and also gave me the expected results in the table:

:mifare: '212821212822212122'
:barcode: 048055048048055049048048049000
:id: '07007100'
:modules: 
id: '07007100'
mifare: '212821212822212122'
barcode: 048055048048055049048048049000
modules: 

If I now remove my the data in the field (to let it solely recreate as the new values) I get this error:

I, [2015-10-27T11:50:38.915151 #29493]  INFO -- : Started POST "/authenticator/hslu/login" for 10.101.83.200 at 2015-10-27 11:50:38 +0000
I, [2015-10-27T11:50:38.916748 #29493]  INFO -- : Processing by Authenticator::HsluAuthenticationController#login as HTML
I, [2015-10-27T11:50:38.916869 #29493]  INFO -- :   Parameters: {"utf8"=>"âœ“", "authenticity_token"=>"RaNYEDonRvY7KHxdYzePLKfbKo4Eau92Wi5rf1uHd3Mex40WXm8GaykRaNZFv3yiM/wwlegxoih49iP7Xl3sEA==", "login"=>{"username"=>"huna1_07", "password"=>"[FILTERED]"}}
D, [2015-10-27T11:50:38.919339 #29493] DEBUG -- :   [1m[35mLanguage Load (0.2ms)[0m  SELECT  `languages`.* FROM `languages` WHERE `languages`.`default` = 1  ORDER BY `languages`.`name` ASC LIMIT 1
D, [2015-10-27T11:50:38.920836 #29493] DEBUG -- :   [1m[36mCACHE (0.0ms)[0m  [1mSELECT  `languages`.* FROM `languages` WHERE `languages`.`default` = 1  ORDER BY `languages`.`name` ASC LIMIT 1[0m  [["default", true]]
D, [2015-10-27T11:50:38.921502 #29493] DEBUG -- :   [1m[35mLanguage Load (0.1ms)[0m  SELECT `languages`.* FROM `languages` WHERE `languages`.`active` = 1  ORDER BY `languages`.`name` ASC
D, [2015-10-27T11:50:38.922108 #29493] DEBUG -- :   [1m[36mCACHE (0.0ms)[0m  [1mSELECT `languages`.* FROM `languages` WHERE `languages`.`active` = 1  ORDER BY `languages`.`name` ASC[0m  [["active", true]]
D, [2015-10-27T11:50:38.993017 #29493] DEBUG -- :   [1m[35mUser Load (0.1ms)[0m  SELECT  `users`.* FROM `users` WHERE `users`.`unique_id` = 'HUNA1_07' LIMIT 1
I, [2015-10-27T11:50:39.007618 #29493]  INFO -- : Completed 500 Internal Server Error in 91ms (ActiveRecord: 1.5ms)
F, [2015-10-27T11:50:39.008463 #29493] FATAL -- : 
NoMethodError (undefined method `[]=' for nil:NilClass):
  app/controllers/authenticator/hslu_authentication_controller.rb:90:in `update_user'
  app/controllers/authenticator/hslu_authentication_controller.rb:174:in `login'



where line 90 is the first line of my new code above. Once the user has logged in the first time I can successfully comment out the original code again.

**Comment by psy-q** [comment]
Ah yes, if `extended_info` hasn't existed at this time it needs to be initialized to an empty hash before you first assign something:
```ruby
user.extended_info = {}
```
From then on you can begin assigning things to keys and values.

**Comment by AnthonyHunt** [comment]
ahhh! ok.  I'll just set that as we wipe out the fields anyway.
